syntax = "proto3";

message mapper_request {
    string file_name = 1;
    repeated string lines = 2;
    string map_function = 3;
}

message tuple {
    string key = 1;
    string value = 2;
}

message tuple_list {
    repeated tuple result =1;
}

message mapper_response {
    repeated tuple result = 1;
}

message address {
    string name = 1;
    string ip = 2;
    int32 port = 3;
}

message connection_response {
    string data = 1;
}

message reducer_response {
    repeated tuple result = 1;
}

message reducer_request {
    string reducer_function = 1;
    repeated tuple result = 2;
}

service Worker {

    rpc worker_map(mapper_request) returns (mapper_response) {}
    rpc connect_to_store(address) returns (connection_response) {}
    rpc worker_reducer(reducer_request) returns (reducer_response) {}
}
